lsp/nls/src/requests/formatting.rs

```diff
@@ -31,7 +31,9 @@ pub fn handle_format_document(
         file: params.text_document.uri,
     })?;
 
-    let result = Some(vec![TextEdit {
+    // TODO: instead of always sending a huge edit, we should compute a diff
+    // between `text` and `formatted` and send more granular edits.
+    let result = (text != formatted).then_some(vec![TextEdit {
```

Comment:
Do you think testing for equality on a full text is necessarily a good tradeoff? The cost is the offset of the first differing character, which can be big if the source is big and the formatting only happen at the end (or is equal). I wonder if we really gain something over sending the whole text anyway...

Reply:
I cribbed that from rust-analyzer and wasn't really thinking of performance, but of not sending edits that do nothing. That prevents, for example, helix from selecting the entire file contents on every format, even if nothing changed.

Reply:
From looking at that very briefly just now, I think we're actually sending a slightly wrong text edit. I'm suspecting there's an off by one error for a trailing newline somewhere, but that's better left for a follow-up PR.

Reply:
IIRC, LSP numbergin starts at 1 for both chars and lines, if that matters. Otherwise, if rust-analyzer is doing it, then it's probably reasonable

Reply:
At least in the completion messages, I'm pretty sure they both start at zero...

Reply:
They do indeed start at zero: https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#range